### PR TITLE
Sum of the squares of the numbers in a list

### PR DESCRIPTION
**Pull request by Mithchell509**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [ ] The function has correct type annotations
- [ ] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_of_squares.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Created 01-11-2025
```
Comment (yuri-spizhovyi-mit) [inline solutions/sum_of_squares.py:5]: The module docstring is present, but it could be more detailed and structured. For example, it should include a description of the module's purpose and its contents, similar to:
```
"""
A module for recursively calculating the sum of digits of an integer.

Module contents:
    - sum_of_digits: A recursive function to calculate the sum of digits.

Created on 2024-12-31
@author: Yurii Spizhovyi
"""
```

File: solutions/sum_of_squares.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    Raises:
+        TypeError: If the input is not a list or contains invalid elements.
+
+    Examples:
+        >>> sum_of_square([1, 2, 3])
+        14
+        >>> sum_of_square([2.5, 1.1, 4])
+        23.46
+        >>> sum_of_square([-5.3, 9, -11, 2])
```
Comment (yuri-spizhovyi-mit) [inline solutions/sum_of_squares.py:33]: There is an issue with the doctest for the input [-5.3, 9, -11, 2]. The expected result is 23.09, but the actual result is 234.09. This discrepancy indicates either an issue with the test case or a potential bug in the implementation of the sum_of_square function. Please review the logic of the function and ensure it calculates the correct sum of squares as described in the docstring. Additionally, verify the precision handling for floating-point numbers to ensure it aligns with the test case expectations.
image

**Review by yuri-spizhovyi-mit — COMMENTED** [review]
Overall, the implementation is solid and well-structured. However, please take a look at the two comments regarding the module docstring and the failed doctest to address the identified issues.

**Comment by Mithchell509** [comment]
Thank You for your comments

**Review by yuri-spizhovyi-mit — APPROVED** [review]
Great work on addressing the feedback and making the necessary changes! The update to the sum_of_square function looks solid, and everything works as expected. Nicely done! 🎉

**Review by Khadijaramlawi — APPROVED** [review]
Great work Mithchell!
